### ISSUE TITLE
Hindsight experience transformer warns incorrectly

### ISSUE THREAD

**Issue by JakobThumm** [HorizonRobotics/alf#1118]
Hi,

I'm using HER to train the `fetchpush` task.
I get a very long warning message at every training step.
The message is created here:
https://github.com/HorizonRobotics/alf/blob/ee1e020ee732468d2f94629b0ad29feb68090075/alf/algorithms/data_transformer.py#L823-L835
If I understand this correctly, the condition checks if not all new rewards are the same as the old rewards.
Since the task of HER is to create new rewards based on new goals, this check should be exactly the opposite
```
if torch.allclose(relabeled_rewards[non_her_indices],
                              result.reward[non_her_indices]):
 ** WARN **
```
instead of 
```
if not torch.allclose(relabeled_rewards[non_her_indices],
                              result.reward[non_her_indices]):
 ** WARN **
```

The bug should be reproducible on the mujoco docker using [this gin file](https://github.com/HorizonRobotics/alf/blob/pytorch/alf/examples/her_fetchpush.

Did I get this correctly?

Best,

Jakob

**Comment by le-horizon** [comment]
Hey @JakobThumm,

This check is a sanity check to make sure the reward function used during training is exactly the same as the reward function used by the environment.  That's why the check is performed on the non_her slice of the data, and checks that they are all close to the environment rewards.

I've looked into it before, and unfortunately the difference exists despite the two reward functions looking exactly the same.

The good thing is, it only happens infrequently and does not affect performance.

Sorry for the confusion.  If it wastes too much disc space for you, you can maybe decrease the logging frequency.

**Comment by hnyu** [comment]
> Hey @JakobThumm,
> 
> This check is a sanity check to make sure the reward function used during training is exactly the same as the reward function used by the environment. That's why the check is performed on the non_her slice of the data, and checks that they are all close to the environment rewards.
> 
> I've looked into it before, and unfortunately the difference exists despite the two reward functions looking exactly the same.
> 
> The good thing is, it only happens infrequently and does not affect performance.
> 
> Sorry for the confusion. If it wastes too much disc space for you, you can maybe decrease the logging frequency.

If it's because of floating errors, maybe we can specify a tolerance to torch.allclose? 
``` python
torch.allclose(input, other, rtol=1e-05, atol=1e-08, equal_nan=False)
```

**Comment by le-horizon** [comment]
Good guess, but it's not because of that, for example, in the case below, clearly the environment rewards are wrong, relabeled rewards are correct (ag not within 0.05 of dg):

relabeled_reward
tensor([-1., -1., -1.,  ..., -1., -1., -1.], device='cpu')
!=
env_reward
tensor([0., 0., 0.,  ..., 0., 0., 0.], device='cpu')
ag:
tensor([[1.4633, 0.7672, 0.4247],
        [1.4633, 0.7672, 0.4247],
        [1.4633, 0.7672, 0.4247],
        ...,
        [1.4935, 0.7783, 0.4247],
        [1.4935, 0.7783, 0.4247],
        [1.4935, 0.7783, 0.4247]], device='cpu')
dg:
tensor([[1.2638, 0.6421, 0.4247],
        [1.2638, 0.6421, 0.4247],
        [1.2638, 0.6421, 0.4247],
        ...,
        [1.4280, 0.6657, 0.4247],
        [1.4280, 0.6657, 0.4247],
        [1.4280, 0.6657, 0.4247]], device='cpu')

**Comment by hnyu** [comment]
> Good guess, but it's not because of that, for example, in the case below, clearly the environment rewards are wrong, relabeled rewards are correct (ag not within 0.05 of dg):
> 
> relabeled_reward tensor([-1., -1., -1., ..., -1., -1., -1.], device='cpu') != env_reward tensor([0., 0., 0., ..., 0., 0., 0.], device='cpu') ag: tensor([[1.4633, 0.7672, 0.4247], [1.4633, 0.7672, 0.4247], [1.4633, 0.7672, 0.4247], ..., [1.4935, 0.7783, 0.4247], [1.4935, 0.7783, 0.4247], [1.4935, 0.7783, 0.4247]], device='cpu') dg: tensor([[1.2638, 0.6421, 0.4247], [1.2638, 0.6421, 0.4247], [1.2638, 0.6421, 0.4247], ..., [1.4280, 0.6657, 0.4247], [1.4280, 0.6657, 0.4247], [1.4280, 0.6657, 0.4247]], device='cpu')

Then it seems to me that this is a bug which needs to be addressed?

**Comment by le-horizon** [comment]
Turns out the checking includes FIRST time steps, which always have reward 0 from the env.
#1120 fixes the aggressive checking.  It should not impact performance.

**Comment by JakobThumm** [comment]
Hi,
That seems to fix the issue. :+1: 
However, I was wondering, is there any particular reason why you give the `reward_fn` as an argument and not use the `compute_reward` reward function from `gym.core.GoalEnv`. 
https://github.com/openai/gym/blob/b84b69c872a3159900e6ec82a4b98cfa3e7bb0ed/gym/core.py#L190-L209

It seems like this function was particularly designed for that. Stable baselines 3 uses that function to compute the reward.
One could ofc define the reward function in the gin config (or py config), however, I see two problems here:
1. It is not obvious that this must be done. Maybe we can change the default to using the normal `compute_reward` function of the environment instead of the `l2_dist_close_reward_fn`?
2. The function call in https://github.com/HorizonRobotics/alf/blob/ee1e020ee732468d2f94629b0ad29feb68090075/alf/algorithms/data_transformer.py#L813 is missing the info field.

EDIT: I see two problems in changing this:
1. You cannot easily pass an environment function to the constructor due to the modular design of alf. This would need some significant changes in the constructor call of data transformers, which is certainly not intended.
2. The `env_info` is not tracked in `experience` https://github.com/HorizonRobotics/alf/blob/ee1e020ee732468d2f94629b0ad29feb68090075/alf/algorithms/rl_algorithm.py#L495-L505

EDIT: I created a branch, where I added `env_info` to `Experience`. However, I think I cannot contribute to this repo due to access rights. I added my `git diff` as a file. Maybe that helps :)
[git_diff_b8b38e866c80c2e0b875aa29985ce583d47d8557.log](https://github.com/HorizonRobotics/alf/files/7668472/git_diff_b8b38e866c80c2e0b875aa29985ce583d47d8557.log)

**Comment by le-horizon** [comment]
@JakobThumm Thank your for your thoughts of improvement and code.  You are always welcome to contribute by creating pull requests.  We'll check whether that permission is there.

You are correct in all your points.  There are probably easy ways to make it use the reward fn of the environment.

Re: Adding env_info to experience: Once inside experience, it will be stored into the replay buffer.  If that's the intention, then you can do that for that env using e.g. gym_wrappers.  However, for generality, some other env's may return lots of env_info e.g. for debugging, so it's not always advisable to store env_info into experience.

**Comment by JakobThumm** [comment]
> There are probably easy ways to make it use the reward fn of the environment.

As I'm just starting using alf, I found that quite hard. It seems to me that the data transformer has no access to the env.

> Once inside experience, it will be stored into the replay buffer. If that's the intention, then you can do that for that env using e.g. gym_wrappers. However, for generality, some other env's may return lots of env_info e.g. for debugging, so it's not always advisable to store env_info into experience.

Yes, that was the intention. It makes sense to not add that to Experience in the general case. You are right about that. I currently don't see how I could write a wrapper that has access to the replay buffer, any tips on where to start?

Thanks for the help :)

**Comment by hnyu** [comment]
> > There are probably easy ways to make it use the reward fn of the environment.
> 
> As I'm just starting using alf, I found that quite hard. It seems to me that the data transformer has no access to the env.
> 
> > Once inside experience, it will be stored into the replay buffer. If that's the intention, then you can do that for that env using e.g. gym_wrappers. However, for generality, some other env's may return lots of env_info e.g. for debugging, so it's not always advisable to store env_info into experience.
> 
> Yes, that was the intention. It makes sense to not add that to Experience in the general case. You are right about that. I currently don't see how I could write a wrapper that has access to the replay buffer, any tips on where to start?
> 
> Thanks for the help :)

No sure if I understand your question clearly, but in general in ALF it's not possible for a gym wrapper to get access to the replay buffer, because the former belongs to the environment and the latter belongs to the algorithm. 

If you really want to use the reward function provided by Gym, then one simple way is to directly import that function into your py config file and pass it as the current `reward_fn` argument. Note that as I checked, even though Gym's `compute_reward` has an additional arg `info`, they never used it. 

But generally I don't think Gym's current way of putting hindsight reward computation into the environment is a good design because this part should belong to an algorithm. You can imagine that if we want to have new hindsight reward computed in a different way, we certainly don't want to touch the environment code.

**Comment by le-horizon** [comment]
> > There are probably easy ways to make it use the reward fn of the environment.
> 
> As I'm just starting using alf, I found that quite hard. It seems to me that the data transformer has no access to the env.

Haonan (hnyu) suggested a good way there.

> > Once inside experience, it will be stored into the replay buffer. If that's the intention, then you can do that for that env using e.g. gym_wrappers. However, for generality, some other env's may return lots of env_info e.g. for debugging, so it's not always advisable to store env_info into experience.
> 
> Yes, that was the intention. It makes sense to not add that to Experience in the general case. You are right about that. I currently don't see how I could write a wrapper that has access to the replay buffer, any tips on where to start?
> 
> Thanks for the help :)

What I mean is if we want to use some fields from env_info during training, we should perhaps put it as part of observation?

Some example environment wrappers:
https://github.com/HorizonRobotics/alf/blob/04a4fde6805fd27bbda344453de90c70ea7a363e/alf/environments/suite_robotics.py#L62

Cheers,
Le

**Comment by JakobThumm** [comment]
> What I mean is if we want to use some fields from env_info during training, we should perhaps put it as part of observation?

That could be a solution. It is not according to the OpenAI gym structure but certainly works.

> If you really want to use the reward function provided by Gym, then one simple way is to directly import that function into your py config file and pass it as the current reward_fn argument.

That works iff it is not a member of the environment class. Because passing `compute_reward(self, achieved_goal, desired_goal, info)` would require an `env` object to get `env.compute_reward`, which is not yet created at the stage of the config file.

I now hacked a version together that works with my environment and I'm testing the functionality :)

Thanks for your help.

**Comment by le-horizon** [comment]
@JakobThumm, glad you made it work.

Regarding passing compute_reward to replay_buffer, with gin config, it's perhaps impossible, but with alf config, you could.
You can call `env = alf.get_env()` inside alf/examples/benchmarks/fetch/fetch_conf.py, after env setup.
I haven't tried it myself, but looks viable.  Cheers.
